File: dotnet/base/Mcma.Api/Routing/Defaults/DefaultRouteCollectionBuilder.cs
using System;
using System.Linq;
using System.Linq.Expressions;
using System.Reflection;
using Mcma.Api.QueryFilters;
using Mcma.Api.Routes;
using Mcma.Api.Routing.Defaults.Routes;
using Mcma.Core;
using Mcma.Core.Utility;
using Mcma.Data;

namespace Mcma.Api.Routing.Defaults
{
    public class DefaultRouteCollectionBuilder<TResource> where TResource : McmaResource
    {
        public DefaultRouteCollectionBuilder(IDbTableProvider dbTableProvider, string root = null)
        {
            DbTableProvider = dbTableProvider;

            root = root ?? typeof(TResource).Name.CamelCaseToKebabCase().PluralizeKebabCase();
            if (!root.StartsWith("/"))
                root = "/" + root;

            Root = root;
            
            Routes = new DefaultRouteCollection<TResource>
            {
                Query = DefaultQueryBuilder.Get<TResource>(DbTableProvider, new InMemoryQueryFilterExpressionProvider(), Root),
                Create = DefaultCreateBuilder.Get<TResource>(DbTableProvider, Root),
                Get = DefaultGetBuilder.Get<TResource>(DbTableProvider, Root),
                Update = DefaultUpdateBuilder.Get<TResource>(DbTableProvider, Root),
                Delete = DefaultDeleteBuilder.Get<TResource>(DbTableProvider, Root)
            };
        }

        private IDbTableProvider DbTableProvider { get; }

        private string Root { get; }

        private DefaultRouteCollection<TResource> Routes { get; }

        public McmaApiRouteCollection Build()
            => new McmaApiRouteCollection(Routes.Included.Select(rb => rb.Build()));

        public DefaultRouteCollectionBuilder<TResource> AddAll()
        {
            Routes.AddAll();
            return this;
        }

        public DefaultRouteConfigurator<TResource, TResult> Route<TResult>(
            Expression<Func<DefaultRouteCollection<TResource>, DefaultRouteBuilder<TResult>>> selectRoute)
        {
            if (!(selectRoute.Body is MemberExpression memberExpression) ||
                !(memberExpression.Member is PropertyInfo propertyInfo) ||
                !typeof(DefaultRouteBuilder<TResult>).IsAssignableFrom(propertyInfo.PropertyType))
                throw new Exception($"Invalid route selection expression: {selectRoute}");

            var routeBuilder = selectRoute.Compile().Invoke(Routes);
            if (routeBuilder == null)
                throw new Exception("Invalid route selection expression");

            return new DefaultRouteConfigurator<TResource, TResult>(this, Routes, routeBuilder);
        }
    }
}

File: nodejs/base/api/lib/routing/defaults/default-route-collection-builder.ts
import { Utils, McmaResourceType, McmaResource, McmaException } from "@mcma/core";
import { DocumentDatabaseTableProvider } from "@mcma/data";

import { camelCaseToKebabCase, pluralizeKebabCase } from "../../strings";
import { McmaApiRouteCollection } from "../route-collection";

import * as routes from "./routes";

import { DefaultRouteConfigurator } from "./default-route-configurator";
import { DefaultRoutes } from "./default-routes";
import { DefaultRouteCollection } from "./default-route-collection";
import { DefaultRouteBuilder } from "./default-route-builder";

export class DefaultRouteCollectionBuilder<T extends McmaResource> {
    private routes: DefaultRouteCollection<T>;

    constructor(
        dbTableProvider: DocumentDatabaseTableProvider,
        resourceType: McmaResourceType<T>,
        root?: string
    ) {
        resourceType = Utils.getTypeName(resourceType);
        if (!resourceType) {
            throw new McmaException("Invalid resource type specified for default routes.");
        }

        root = root || pluralizeKebabCase(camelCaseToKebabCase(resourceType));
        if (root[0] !== "/") {
            root = "/" + root;
        }

        this.routes = new DefaultRouteCollection<T>({
            query: routes.defaultQueryBuilder<T>(dbTableProvider, root),
            create: routes.defaultCreateBuilder<T>(dbTableProvider, root),
            get: routes.defaultGetBuilder<T>(dbTableProvider, root),
            update: routes.defaultUpdateBuilder<T>(dbTableProvider, root),
            delete: routes.defaultDeleteBuilder<T>(dbTableProvider, root)
        });
    }

    build(): McmaApiRouteCollection {
        return new McmaApiRouteCollection(this.routes.included.map(rb => rb.build()));
    }

    addAll(): this {
        this.routes.addAll();
        return this;
    }

    route<TResult>(selectRoute: (defaultRoutes: DefaultRoutes<T>) => DefaultRouteBuilder<TResult>): DefaultRouteConfigurator<T, TResult> {
        const routeBuilder = selectRoute(this.routes);
        if (!routeBuilder) {
            throw new McmaException("Invalid route selection expression");
        }
        return new DefaultRouteConfigurator<T, TResult>(this, this.routes, routeBuilder);
    }
}
